architecture/verifier/freshness.md

```diff
@@ -0,0 +1,101 @@
+freshness.
+
+![nonce flow diagram](nonce-sequence.png)
+
+1. Verification is initiated by the client (the Attester or the Relying Party,
+   depending on the [interaction
+   model](https://www.rfc-editor.org/rfc/rfc9334.html#name-topological-patterns)
+   used) with a new session request. The client can optionally specify the
+   nonce to be used for this verification session, or the size of the nonce to
+   be generated by Veraison. Either way, the nonce is limited to be between 8
+   and 64 bytes long (see [EAT eat_nonce
+   claim](https://www.ietf.org/archive/id/draft-ietf-rats-eat-21.html#name-eat_nonce-eat-nonce-claim)).
+2. The nonce (whether provided or generated) is associated with the created
+   session, and is returned in the response to the new session request.
+3. (The assumption is that) the returned nonce is then used as the challenge to
+   the attester when generating the attestation token.
+4. The attestation token/evidence (that embeds the nonce/challenge) is then
+   submitted to the session URL for verification.
+5. The verification frontend retrieves the nonce it has associated with the
+   session and attaches it to the received attestation token, and sends to the
+   verifier on the VTS backend via a `IVTSClient.GetAttestation()` call.
+6. The VTS verifier uses media type information associated with the attestation token
+   to retrieve an appropriate handler from the loaded plugins, and uses that
+   handler to process the evidence.
+7. That includes invoking `IEvidenceHandler.ValidateEvidenceIntegrity()`, which
+   is intended for the handler implementation to provide scheme-specific
+   means of making sure the received evidence is valid. This is the place at
+   which handler implementations (i.e. scheme plugins) MAY check that the
+   session nonce associated with the token matches the challenge decoded from
+   the token's evidence claims.
+8. Once the scheme-specific processing of attestation evidence is complete, the
+   results a passed to the policy agent via a call to `IAgent.Evaluate()`. This
+   allows polices an opportunity to override the results generated by the
+   scheme handler, including downgrading result status if the session nonce
+   does not match the challenge inside the evidence.
+9. Finally, the attestation result is signed and returned to the client. The
+   session nonce and the extracted evidence are part of the result, allowing
+   for the final opportunity to confirm freshness by the receiver.
+
+
+### Checking via scheme handler
+
+Freshness can be verified inside `IEvidenceHandler.ValidateEvidenceIntegrity()`
+implementation. This receives a `proto.AttestationToken` as an argument. The
+session nonce is contained within `token.Nonce` field, and the `token.Evidence`
+field contains the raw bytes of the evidence set by the client. See
+[PSA_IOT](https://github.com/veraison/services/blob/main/scheme/psa-iot/evidence_handler.go#L78)
```

Comment:
question: (for reference stability reasons) would it be possible to link the name rather than the line here?

Reply:
Good question. I don't belive so, since that would require GitHub to resolve symbols, which I don't believe it does. Will have a look though.

Reply:
Right, doesn't seem so, however i've changed the ref part of the URL path from `main` to the specific hash of the current commit, so should be stable.

(see also: https://stackoverflow.com/questions/18220650/github-link-to-function-in-source)

Reply:
GitHub can resolve symbols (see https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github).  However, looking at the way the URLs for the symbol definition are constructed (i.e., with the `#Ln` fragment) it's probably a dynamic thing...

Reply:
Ah, interseting. TIL.